### ISSUE TITLE
No equivalent atom in ir_interpolate_at with CaHgO2

### ISSUE THREAD

**Issue by rebeccafair** [brille/brille#63]
When trying to interpolate at any q-point with CaHgO2 I get the following error:
```
2 : ( 0.894013 0.894013 0.894013 )
2 : ( 0.105987 0.105987 0.105987 )
0 : ( 0.500000 0.500000 0.500000 )
1 : ( 0.000000 1.000000 0.000000 )

does not have an equivalent atom to 0 for symmetry operation
  0 -1  0 -1  0  0  0  0 -1+  0.0000  0.0000  0.0000
Traceback (most recent call last):
  File "brille_bug.py", line 53, in <module>
    grid.ir_interpolate_at(qpts)
RuntimeError: All atoms in the basis *must* be mapped to an equivalent atom for *all* symmetry operations
```
This can be reproduced with the following script:
```
import spglib as spg
import numpy as np
from euphonic import ForceConstants

import brille as br

# Get crystal data
fc = ForceConstants.from_json_file('CaHgO2_force_constants.json')
crystal = fc.crystal
cell_vectors = crystal._cell_vectors
cell = crystal.to_spglib_cell()

# Create grid from crystal
dataset = spg.get_symmetry_dataset(cell)
rotations = dataset['rotations']  # in fractional
translations = dataset['translations']  # in fractional
symmetry = br.Symmetry(rotations, translations)
direct = br.Direct(*cell)
direct.spacegroup = symmetry
bz = br.BrillouinZone(direct.star)
vol = bz.ir_polyhedron.volume
grid = br.BZTrellisQdc(bz, node_volume_fraction=vol/1000)

# Fill grid with whatever
n_atoms = crystal.n_atoms
values = np.ones((len(grid.rlu), 3*n_atoms, 1), dtype=np.float64)
vectors = np.ones((len(grid.rlu), 3*n_atoms, 3*n_atoms), dtype=np.complex128)
grid.fill(values, (1,), (1., 0., 0.),
          vectors, (0., 3*n_atoms, 0, 3, 0, 0), (0., 1., 0.))

# Cause error
qpts = np.array([[0.05    ,  0.05    ,  0.05]])
grid.ir_interpolate_at(qpts)
```

The Euphonic force constants file used in the script can be found [here](https://github.com/pace-neutrons/Euphonic/blob/master/tests_and_analysis/test/data/force_constants/CaHgO2_force_constants.json).

The error seems to happen with any grid type or grid parameters. I'm using brille 0.5.7.

**Comment by ajjackson** [comment]
I'm getting a similar error message with data from http://phonondb.mtl.kyoto-u.ac.jp/ph20180417/d000/mp-661.html
which is quite a small and high-symmetry cell

**Comment by g5t** [comment]
The problem with CaHgO<sub>2</sub> is that the atom positions in your force constant file do not follow the symmetry of the lattice, likely due to rounding errors or truncation. 

The symmetry operations of require that all atoms have equivalent positions under exchange of any two coordinates or under space-inverseion.
If we take the [coordinates of the first oxygen atom ](https://github.com/pace-neutrons/Euphonic/blob/aa3cc28786797bb3052f898dd63d4928d6f27ee2/tests_and_analysis/test/data/force_constants/CaHgO2_force_constants.json#L149-L151) as `(a b a)` and the [coordinates of the second oxygen atom](https://github.com/pace-neutrons/Euphonic/blob/aa3cc28786797bb3052f898dd63d4928d6f27ee2/tests_and_analysis/test/data/force_constants/CaHgO2_force_constants.json#L154-L156) as `(c d c) == (1-a 1-b 1-a)` we see that space-inversion symmetry is fulfilled for the oxygens, but the system is missing atoms at `(b a a)`, `(a a b)`, `(1-b 1-a 1-a)`, and `(1-a 1-a 1-b)` for the remaining symmetry elements.
The difference between `a` and `b` is only 10<sup>-8</sup> Å, so it seems reasonable that the oxygen positons should be interpreted as being along `[111]`, in which case all symmetry operations would be fulfilled. Unfortunately `brille <= v0.5.7` uses a hard-coded floating-point approximate comparison intended only to account for rounding errors which has a tolerance much less than 10<sup>-8</sup>, so `brille` can only treat the specified atom basis as complete with modifications.

Issue #64 has its origins in `brille`'s floating point comparison tolerance as well, so the necessary modifications for both issues have been carried out in a [common branch](https://github.com/brille/brille/tree/64_polyhedra).
The modifications in that branch have introduced a breaking API change, so once again the 'minor' pre-release version number has been incremented to `v0.6.0`. 
The script above, modified to work with the development `v0.6.0` is 

```python
import spglib as spg
import numpy as np
from euphonic import ForceConstants

import brille as br
print(f"Using brille version {br.version} built {br.build_datetime} or {br.build_hostname}")

# The default floating point comparison tolerances are too small
# The biggest offender is Hg, which at [0, 0.99999999, 0] needs a tolerance in
# excess of 2e-8 due to the matching-mechanism within brille.
# But the reciprocal-space tolerances can be tighter, 5e-10 seems to work.
print(f"Module-global floating point tolerances {br.real_space_tolerance()=} and {br.reciprocal_space_tolerance()=}")
# With v0.6.0 we can now change the configuration as a local object:
ac = br.ApproxConfig()
ac.real_space_tolerance = 2.1e-8
ac.reciprocal_space_tolerance = 5e-10
# # or by modifying the module-wide object (which uses methods instead of properties)
# br.real_space_tolerance(2.1e-8)
# br.reciprocal_space_tolerance(5e-10)


# Get crystal data
euphonic_root = 'path_to_your_euphonic_repository'
filename=f"{euphonic_root}/tests_and_analysis/test/data/force_constants/CaHgO2_force_constants.json"
fc = ForceConstants.from_json_file(filename)
crystal = fc.crystal
cell_vectors = crystal._cell_vectors
basis_vectors, atom_positions, atom_types = crystal.to_spglib_cell()

# Create grid from crystal
dataset = spg.get_symmetry_dataset((basis_vectors, atom_positions, atom_types))
rotations = dataset['rotations']  # in fractional
translations = dataset['translations']  # in fractional

symmetry = br.Symmetry(rotations, translations)
basis = br.Basis(atom_positions, atom_types)
lattice = br.Lattice(basis_vectors, symmetry, basis)  # holds both the Direct and Reciprocal lattice parameters

bz = br.BrillouinZone(lattice)
vol = bz.ir_polyhedron.volume
# grid = br.BZTrellisQdc(bz, node_volume_fraction=vol/30)  # use with module-wide tolerances
grid = br.BZTrellisQdc(bz, node_volume_fraction=vol/30, always_triangulate=False, approx_config=ac)  # use with local tolerances

# Fill grid with whatever
n_atoms = crystal.n_atoms
values = np.ones((len(grid.rlu), 3*n_atoms, 1), dtype=np.float64)
vectors = np.ones((len(grid.rlu), 3*n_atoms, 3*n_atoms), dtype=np.complex128)
grid.fill(values, (1,), (1., 0., 0.),
          vectors, (0., 3*n_atoms, 0, 3, 0, 0), (0., 1., 0.))

# [Don't] Cause error
qpts = np.array([[0.05    ,  0.05    ,  0.05]])
grid.ir_interpolate_at(qpts)
```

With the new `brille.ApproxConfig` object, it's now up to the user to decide what tolerances are appropriate for their system.
The atomic positions in [AlN](http://phonondb.mtl.kyoto-u.ac.jp/ph20180417/d000/mp-661.html#poscar-yaml) seem to suffer from the same problem and, I suspect but have not checked that, a similar `brille.ApproxConfig.real_space_tolerance` would fix the error message encountered by @ajjackson.

**Comment by rebeccafair** [comment]
Thanks, the above script seems to work for me with Brille `0.6.0+64_polyhedra.8a5e72b`.

I've been playing around with the tolerances and a few things have come up:

* If I set `ac.reciprocal_space_tolerance = 1e-5` and `ac.reciprocal_space_tolerance = 2.1e-8`  the call to `BZTrellisQdc` just hangs forever.
* If it set both `ac.reciprocal_space_tolerance = 1e-5` and `ac.reciprocal_space_tolerance = 1e-5` and include `ac` in the BZ creation `bz = br.BrillouinZone(lattice, approx_config=ac)` and try using it with some random q-points I get the following error:
```
The node subscript The node subscript The node subscript   0,  4, 16, for the point (hkl) [[-1.1787743726731688, -0.0090823494721997,  0.1408328076262703]](xyz)   6, 14,  7, for the point (hkl) [[-0.9506316474702963,  0.4090050856308074, -0.2234854549655343]](xyz) The node subscript  27,  2, 19, for the point (hkl) [[-0.1394477661082921, -0.0496127626012558,  0.2631071318179671]](xyz)  -1.1787743726731688, -0.0090823494721997,  0.1408328076262703, is either invalid or points to a null node!  6, 16, 10, for the point (hkl) [[-0.9431210364514688,  0.4773902500911960, -0.0903758182591777]](xyz)  -0.1394477661082921, -0.0496127626012558,  0.2631071318179671, is either invalid or points to a null node!
 -0.943
```
But if I create the BZ without `ac` it works ok. As long as I'm consistent and providing the same `ac` to both `BrillouinZone` and `BZTrellisQdc` shouldn't it work? 

Ideally, there would be a general solution or way to figure out what the settings should be, rather than having to do trial and error, e.g. with LZO I have to provide `ac` to `BrillouinZone` only, and with CaHgO2 it seems like I have to provide `ac` to `BZTrellisQdc` only, and changing the tolerances seems to sometimes have unexpected results (like the hanging in the above example). I feel like users might need a lot of hand holding to find the things that work for their specific case (like the hand holding I'm needing right now!). Or is a general solution not possible?

**Comment by g5t** [comment]
@rebeccafair Can you check whether the newest `brille` module on PyPI still produces the infinite hang and error message you described above. I've tried reproducing both running on a Linux and macOS machine and not been able to reproduce the hang or error with Python 3.10 and brille==0.6.0.

**Comment by g5t** [comment]
This issue was reopened due to two problems: 1) a reproducible infinite-hang; and 2) problems with setting appropriate floating point comparison tolerances.

1. The first issue was caused by a bug, fixed in #79.
2. The second has hopefully been addressed in #77 

With both issues closed, this will be close as well. If either re-appears with the newest changes we can revisit the issue.
